lib/providers/badge_message_provider.dart

```diff
@@ -34,25 +41,61 @@ class BadgeMessageProvider extends ChangeNotifier {
     8: Speed.eight,
   };
 
-  void generateMessage(
+  Data generateData(
       String text, bool flash, bool marq, Speed speed, Mode mode) {
     Data data = Data(messages: [
       Message(
-          text: Converters.messageTohex(text),
-          flash: flash,
-          marquee: marq,
-          speed: speed,
-          mode: mode)
+        text: Converters.messageTohex(text),
+        flash: flash,
+        marquee: marq,
+        speed: speed,
+        mode: mode,
+      )
     ]);
-    dataFormed(data);
-    transferData(data);
+    logger.d(
+        "${data.messages.length} message : ${data.messages[0].text} Flash : ${data.messages[0].flash} Marquee : ${data.messages[0].marquee} Mode : ${data.messages[0].mode}");
+    return data;
   }
 
-  void transferData(Data data) {
-    BadgeMagicBluetooth.scanAndConnect(data);
-    logger.d(".......Data is being transferred.......");
+  Future<void> transferData() async {
+    DateTime now = DateTime.now();
+    BleState state = ScanState();
+    while (state is! CompletedState) {
+      BleState? nextState = await state.processState();
+      if (nextState != null) {
+        state = nextState;
+      } else {
+        break;
+      }
+    }
+    if (state is CompletedState) {
+      await state.processState(); // Ensure the toast is shown
+    }
+    logger.d("Time to transfer data is = ${DateTime.now().difference(now)}");
+    logger.d(".......Data transfer completed.......");
   }
 
-  void dataFormed(Data data) => logger.d(
-      "${data.messages.length} message : ${data.messages[0].text} Flash : ${data.messages[0].flash} Marquee : ${data.messages[0].marquee} Mode : ${data.messages[0].mode}");
+  Future<void> checkAndTransfer() async {
+    if (await FlutterBluePlus.isSupported == false) {
+      toast.failureToast('Bluetooth is not supported by the device');

```

Comment:
```suggestion
      toast.showErrorToast('Bluetooth is not supported by the device');
```

This can be done for all error toasts